services/dynamic-sidecar/src/simcore_service_dynamic_sidecar/utils.py

```diff
@@ -0,0 +1,282 @@
+    with the following syntax `%%container_name.SERVICE_KEY_NAME%%`,
+    where `SERVICE_KEY_NAME` targets a container in the compose spec
+
+    If the directive cannot be applied it will just be left untouched
+    """
+    matches = set(_extract_templated_entries(stringified_compose_spec))
+    for match in matches:
+        parts = match.split(".")
+
+        if len(parts) != 2:
+            continue  # templating will be skipped
+
+        target_property = parts[0]
+        services_key = parts[1]
+        if target_property != "container_name":
+            continue  # also ignore if the container_name is not the directive to replace
+
+        remapped_service_key = spec_services_to_container_name[services_key]
+        replace_with = services.get(remapped_service_key, {}).get(
+            "container_name", None
+        )
+        if replace_with is None:
+            continue  # also skip here if nothing was found
+
+        match_pattern = f"%%{match}%%"
+        stringified_compose_spec = stringified_compose_spec.replace(
+            match_pattern, replace_with
+        )
+
+    return stringified_compose_spec
+
+
+def _merge_env_vars(
+    compose_spec_env_vars: List[str], settings_env_vars: List[str]
+) -> List[str]:
+    def _gen_parts_env_vars(
+        env_vars: List[str],
+    ) -> Generator[Tuple[str, str], None, None]:
+        for env_var in env_vars:
+            key, value = env_var.split("=")
+            yield key, value
+
+    # pylint: disable=unnecessary-comprehension
+    dict_spec_env_vars = {k: v for k, v in _gen_parts_env_vars(compose_spec_env_vars)}
+    dict_settings_env_vars = {k: v for k, v in _gen_parts_env_vars(settings_env_vars)}
+
+    # overwrite spec vars with vars from settings
+    for key, value in dict_settings_env_vars.items():
+        dict_spec_env_vars[key] = value
+
+    # returns a single list of vars
+    return [f"{k}={v}" for k, v in dict_spec_env_vars.items()]
+
+
+def _inject_backend_networking(
+    parsed_compose_spec: Dict[str, Any], network_name: str = "__backend__"
+) -> None:
+    """
+    Put all containers in the compose spec in the same network.
+    The `network_name` must only be unique inside the user defined spec;
+    docker-compose will add some prefix to it.
+    """
+
+    networks = parsed_compose_spec.get("networks", {})
+    networks[network_name] = None
+
+    for service_content in parsed_compose_spec["services"].values():
+        service_networks = service_content.get("networks", [])
+        service_networks.append(network_name)
+        service_content["networks"] = service_networks
+
+    parsed_compose_spec["networks"] = networks
+
+
+def validate_compose_spec(
+    settings: DynamicSidecarSettings, compose_file_content: str
+) -> str:
+    """
+    Checks the following:
+    - proper yaml format
+    - no "container_name" service property allowed, because it can
+        spawn 2 cotainers with the same name
+    """
```

Comment:
why not use here [docker-compose config ... ](https://docs.docker.com/compose/reference/config/)?? I recall creating fixture to compose and validate docker-compose files and was a nightmare ... and the worst is that it changes with the version so it is really difficult to maintain

Reply:
I am sorry this comment is out of date and it no longer applies.